db/src/cache.rs

```diff
@@ -557,8 +1084,361 @@ impl SQLiteAttributeCache {
+        // and then shim it into `destination.`
+        // We haven't implemented this because it does not currently occur.
+        // TODO: do this! Then do this:
+        // destination.inner = Arc::new(self.in_progress);
+        unimplemented!();
+    }
+}
+
+pub struct InProgressCacheTransactWatcher<'a> {
+    // A transaction might involve attributes that we cache. Track those values here so that
+    // we can update the cache after we commit the transaction.
+    collected_assertions: BTreeMap<Entid, Either<(), Vec<(Entid, TypedValue)>>>,
+    collected_retractions: BTreeMap<Entid, Either<(), Vec<(Entid, TypedValue)>>>,
+    cache: &'a mut InProgressSQLiteAttributeCache,
+    active: bool,
+}
+
+impl<'a> InProgressCacheTransactWatcher<'a> {
+    fn new(cache: &'a mut InProgressSQLiteAttributeCache) -> InProgressCacheTransactWatcher<'a> {
+        let mut w = InProgressCacheTransactWatcher {
+            collected_assertions: Default::default(),
+            collected_retractions: Default::default(),
+            cache: cache,
+            active: true,
+        };
+
+        // This won't change during a transact.
+        w.active = w.cache.has_cached_attributes();
+        w
+    }
+}
+
+impl<'a> TransactWatcher for InProgressCacheTransactWatcher<'a> {
+    fn datom(&mut self, op: OpType, e: Entid, a: Entid, v: &TypedValue) {
+        if !self.active {
+            return;
+        }
+
+        let target = if op == OpType::Add {
+            &mut self.collected_assertions
+        } else {
+            &mut self.collected_retractions
+        };
+        match target.entry(a) {
+            Entry::Vacant(entry) => {
+                let is_cached = self.cache.is_attribute_cached_forward(a) ||
+                                self.cache.is_attribute_cached_reverse(a);
+                if is_cached {
+                    entry.insert(Either::Right(vec![(e, v.clone())]));
+                } else {
+                    entry.insert(Either::Left(()));
+                }
+            },
+            Entry::Occupied(mut entry) => {
+                match entry.get_mut() {
+                    &mut Either::Left(_) => {
+                        // Nothing to do.
+                    },
+                    &mut Either::Right(ref mut vec) => {
+                        vec.push((e, v.clone()));
+                    },
+                }
+            },
+        }
+    }
+
+    fn done(&mut self, schema: &Schema) -> Result<()> {
+        // Oh, I wish we had impl trait. Without it we have a six-line type signature if we
+        // try to break this out as a helper function.
+        let collected_retractions = mem::replace(&mut self.collected_retractions, Default::default());
+        let collected_assertions = mem::replace(&mut self.collected_assertions, Default::default());
+        let mut do_twice = once(collected_retractions).chain(once(collected_assertions))
```

Comment:
nit: I hate this var name. It should describe what the var contains, not how many times you should call it.

Reply:
The name is intended to point out that it contains an iterator of two processed things, simply because we have no good way of describing the types (see the comment right above); we use this pattern quite often in the core because simple chaining of iterators this way doesn't allocate. I'll change the name to something else.